File: sdks/go/client_test.go
package hsm

import (
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
)

func TestNewClient(t *testing.T) {
	client := NewClient(ClientConfig{
		BaseURL:      "https://hsm.example.com",
		SessionID:    "test-session",
		SessionToken: "test-token",
	})

	assert.NotNil(t, client)
	assert.Equal(t, "https://hsm.example.com", client.baseURL)
}

func TestClientTrimsTrailingSlash(t *testing.T) {
	client := NewClient(ClientConfig{
		BaseURL: "https://hsm.example.com/",
	})

	assert.Equal(t, "https://hsm.example.com", client.baseURL)
}

func TestIsAuthenticated(t *testing.T) {
	client := NewClient(ClientConfig{
		BaseURL:      "https://hsm.example.com",
		SessionID:    "test-session",
		SessionToken: "test-token",
	})

	assert.True(t, client.IsAuthenticated())
}

func TestIsNotAuthenticated(t *testing.T) {
	client := NewClient(ClientConfig{
		BaseURL: "https://hsm.example.com",
	})

	assert.False(t, client.IsAuthenticated())
}

func TestClearCredentials(t *testing.T) {
	client := NewClient(ClientConfig{
		BaseURL:      "https://hsm.example.com",
		SessionID:    "test-session",
		SessionToken: "test-token",
	})

	client.ClearCredentials()
	assert.False(t, client.IsAuthenticated())
}

func TestSetCredentials(t *testing.T) {
	client := NewClient(ClientConfig{
		BaseURL: "https://hsm.example.com",
	})

	client.SetCredentials("new-session", "new-token")
	assert.True(t, client.IsAuthenticated())
}

func TestCircuitBreakerState(t *testing.T) {
	client := NewClient(ClientConfig{
		BaseURL: "https://hsm.example.com",
	})

	assert.Equal(t, "closed", client.CircuitState())
}

func TestOperationCount(t *testing.T) {
	client := NewClient(ClientConfig{
		BaseURL: "https://hsm.example.com",
	})

	assert.Equal(t, 0, client.OperationCount())
}

func TestCryptoToBase64(t *testing.T) {
	data := []byte("Hello")
	result := ToBase64(data)
	assert.Equal(t, "SGVsbG8=", result)
}

func TestCryptoFromBase64(t *testing.T) {
	result, err := FromBase64("SGVsbG8=")
	assert.NoError(t, err)
	assert.Equal(t, []byte("Hello"), result)
}

func TestIsBase64Valid(t *testing.T) {
	assert.True(t, IsBase64("SGVsbG8="))
	assert.True(t, IsBase64(""))
	assert.True(t, IsBase64("YWJj"))
}

func TestIsBase64Invalid(t *testing.T) {
	assert.False(t, IsBase64("Hello"))
	assert.False(t, IsBase64("abc"))
}

func TestAuthenticationError(t *testing.T) {
	err := NewAuthenticationError("Invalid credentials")
	assert.Equal(t, 401, err.StatusCode)
	assert.Equal(t, "AUTHENTICATION_FAILED", err.Code)
	assert.Contains(t, err.Error(), "Invalid credentials")
}

func TestNotFoundError(t *testing.T) {
	err := NewNotFoundError("Key", "key-123")
	assert.Equal(t, 404, err.StatusCode)
	assert.Equal(t, "Key", err.Resource)
	assert.Equal(t, "key-123", err.ResourceID)
	assert.Contains(t, err.Error(), "Key not found: key-123")
}

func TestRateLimitError(t *testing.T) {
	err := NewRateLimitError("Too many requests", 60)
	assert.Equal(t, 429, err.StatusCode)
	assert.Equal(t, 60, err.RetryAfter)
}

func TestNetworkError(t *testing.T) {
	cause := assert.AnError
	err := NewNetworkError("Connection refused", cause)
	assert.Equal(t, "NETWORK_ERROR", err.Code)
	assert.Equal(t, cause, err.Unwrap())
}

func TestDefaultRetryConfig(t *testing.T) {
	config := DefaultRetryConfig()
	assert.Equal(t, 3, config.MaxRetries)
	assert.Equal(t, 100*time.Millisecond, config.BaseDelay)
	assert.Equal(t, 5*time.Second, config.MaxDelay)
	assert.Contains(t, config.RetryOnStatus, 429)
}

File: sdks/typescript/tests/client.test.ts
/**
 * HSM Client Tests
 */

import { beforeEach, describe, expect, it } from 'vitest';
import { createClient, HsmClient } from '../src/client';
import { fromBase64, isBase64, toBase64 } from '../src/crypto';
import { AuthenticationError, NetworkError, NotFoundError, RateLimitError } from '../src/errors';

describe('HsmClient', () => {
  let client: HsmClient;

  beforeEach(() => {
    client = createClient({
      baseUrl: 'https://hsm.example.com',
      sessionId: 'test-session',
      sessionToken: 'test-token',
    });
  });

  describe('createClient', () => {
    it('should create client with config', () => {
      expect(client).toBeInstanceOf(HsmClient);
    });

    it('should strip trailing slash from baseUrl', () => {
      const trailingSlashClient = createClient({
        baseUrl: 'https://hsm.example.com/',
      });
      expect(trailingSlashClient).toBeInstanceOf(HsmClient);
    });
  });

  describe('authentication', () => {
    it('should report authenticated when credentials set', () => {
      expect(client.isAuthenticated()).toBe(true);
    });

    it('should report not authenticated when no credentials', () => {
      const anonymousClient = createClient({
        baseUrl: 'https://hsm.example.com',
      });
      expect(anonymousClient.isAuthenticated()).toBe(false);
    });

    it('should clear credentials', () => {
      client.clearCredentials();
      expect(client.isAuthenticated()).toBe(false);
    });

    it('should set new credentials', () => {
      client.clearCredentials();
      client.setCredentials('new-session', 'new-token');
      expect(client.isAuthenticated()).toBe(true);
    });
  });

  describe('circuit breaker', () => {
    it('should start closed', () => {
      expect(client.getCircuitState()).toBe('closed');
    });

    it('should reset circuit breaker', () => {
      client.resetCircuitBreaker();
      expect(client.getCircuitState()).toBe('closed');
    });
  });

  describe('operation count', () => {
    it('should start at zero', () => {
      expect(client.getOperationCount()).toBe(0);
    });
  });
});

describe('Crypto Utilities', () => {
  describe('toBase64', () => {
    it('should encode bytes to base64', () => {
      const data = new Uint8Array([72, 101, 108, 108, 111]);
      expect(toBase64(data)).toBe('SGVsbG8=');
    });

    it('should encode string to base64', () => {
      expect(toBase64('Hello')).toBe('SGVsbG8=');
    });

    it('should return base64 string as-is', () => {
      expect(toBase64('SGVsbG8=')).toBe('SGVsbG8=');
    });
  });

  describe('fromBase64', () => {
    it('should decode base64 to bytes', () => {
      const bytes = fromBase64('SGVsbG8=');
      expect(Array.from(bytes)).toEqual([72, 101, 108, 108, 111]);
    });
  });

  describe('isBase64', () => {
    it('should return true for valid base64', () => {
      expect(isBase64('SGVsbG8=')).toBe(true);
      expect(isBase64('')).toBe(true);
      expect(isBase64('YWJj')).toBe(true);
    });

    it('should return false for invalid base64', () => {
      expect(isBase64('Hello')).toBe(false);
      expect(isBase64('abc')).toBe(false);
    });
  });
});

describe('Error Classes', () => {
  it('should create AuthenticationError', () => {
    const error = new AuthenticationError('Invalid credentials');
    expect(error.name).toBe('AuthenticationError');
    expect(error.statusCode).toBe(401);
    expect(error.code).toBe('AUTHENTICATION_FAILED');
  });

  it('should create NotFoundError', () => {
    const error = new NotFoundError('Key', 'key-123');
    expect(error.name).toBe('NotFoundError');
    expect(error.message).toBe('Key not found: key-123');
    expect(error.statusCode).toBe(404);
  });

  it('should create RateLimitError with retry after', () => {
    const error = new RateLimitError('Too many requests', { retryAfter: 60 });
    expect(error.name).toBe('RateLimitError');
    expect(error.retryAfter).toBe(60);
    expect(error.statusCode).toBe(429);
  });

  it('should create NetworkError', () => {
    const cause = new Error('Connection refused');
    const error = new NetworkError('Network error', { cause });
    expect(error.name).toBe('NetworkError');
    expect(error.cause).toBe(cause);
  });
});
